File: src/repository/TwoWayDao.java
package repository;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.function.Consumer;

import javax.persistence.EntityManager;
import javax.persistence.EntityManagerFactory;
import javax.persistence.EntityTransaction;
import javax.persistence.Persistence;
import javax.persistence.Query;

import model.TwoWay;
import model.User;

public class TwoWayDao implements Repository<TwoWay, Integer>{
	
	private EntityManagerFactory emf = Persistence.createEntityManagerFactory(
			"jdbc:sqlserver://localhost/Alibaba.odb;integratedSecurity=true;");
	private EntityManager em=emf.createEntityManager();  

	@Override
	public TwoWay findbyId(Integer id) {
		try {
			return em.find(TwoWay.class, id);
		} catch (Exception e) {
			return null;
		}
	}

	@Override
	public List<TwoWay> findByIDs(Collection<Integer> ids) {
		List<TwoWay> users = new ArrayList<TwoWay>();
		try {
			for(int id : ids) {
				users.add(em.find(TwoWay.class, id));
			}
			return users;
		} catch (Exception e) {
			return null;
		}
	}

	@Override
	public List<TwoWay> findAll() {
		Query q = em.createQuery("select * from Two_Way");
		return q.getResultList();
	}

	@Override
	public Boolean deleteByID(Integer id) {
		try {
			executeInsideTransaction(entityManager -> em.remove(em.find(TwoWay.class, id)));
			return true;
		} catch(Exception e) {
			return false;
		}
	}

	@Override
	public Boolean DeleteByIDs(Collection<Integer> ids) {
		try {
			for(int id : ids) {
				executeInsideTransaction(entityManager -> em.remove(em.find(TwoWay.class, id)));
			}
			return true;
		} catch(Exception e) {
			return false;
		}
	}

	@Override
	public TwoWay save(TwoWay E) {
		try {
			executeInsideTransaction(entityManager -> entityManager.persist(E));
			return em.find(TwoWay.class, emf.getPersistenceUnitUtil().getIdentifier(E));
		} catch(Exception e) {
			executeInsideTransaction(entityManager -> entityManager.merge(E));
			return em.find(TwoWay.class, emf.getPersistenceUnitUtil().getIdentifier(E));
		}
	}
	
	private void executeInsideTransaction(Consumer<EntityManager> action) {
        EntityTransaction tx = em.getTransaction();
        try {
            tx.begin();
            action.accept(em);
            tx.commit(); 
        }
        catch (RuntimeException e) {
            tx.rollback();
            throw e;
        }
    }

}

File: repository/TwoWayDao.py
from Repository import RepositoryInterface
from Connection import getConn
from TwoWay import TwoWay
class TwoWayDao(RepositoryInterface):
    def findById(id):
        try:
            c = getConn()
            c.execute('select * from Two_Way where OverseasFlightTransportnumber1=' + str(id))
            i = c.fetchall()
            return TwoWay(i[0][0], i[0][1], i[0][2], i[0][3])
        except:
            print("There was an error")
            return None
    def findByIds(ids):
        try:
            a = []
            c = getConn()
            for id in ids:
                c.execute('select * from Two_Way where OverseasFlightTransportnumber1=' + str(id))
                i = c.fetchall()
                a.append(TwoWay(i[0][0], i[0][1], i[0][2], i[0][3]))
            return a
        except:
            print("There was an error")
            return None
    def findAll():
        try:
            a = []
            c=getConn()
            c.execute('select * from Two_Way')    
            for i in c:
                a.append(TwoWay(i[0], i[1], i[2], i[3]))
            return a
        except:
            print("There was an error")
            return None
    def deleteById(id):
        try:
            c=getConn()
            c.execute('delete from Two_Way where OverseasFlightTransportnumber1=' + str(id))
            c.commit()
            return True
        except:
            return False
    def deleteByIds(ids):
        try:
            c=getConn()
            for id in ids:
                   c.execute('delete from TwoWay where OverseasFlightTransportnumber1=' + str(id))
                   c.commit()
            return True
        except:
            return False
    def save(entity):
        try:
            c=getConn()
            c.execute('select * from TwoWay where OverseasFlightTransportnumber1=' + str(entity.OverseasFlightTransportnumber1))
            i = c.fetchall()
            if len(i) == 0:
                c.execute('insert into TwoWay values(' + str(entity.OverseasFlightTransportnumber1)+',' +str(entity.OverseasFlightTransportnumber2)+','+ str(entity.DomesticFlightTransportnumber1)+','+ str(entity.DomesticFlightTransportnumber2)+')')
                c.commit()
            else :
                c.execute('update TwoWay set OverseasFlightTransportnumber2=' +str(entity.OverseasFlightTransportnumber2)+', DomesticFlightTransportnumber1='+ str(entity.DomesticFlightTransportnumber1)+', DomesticFlightTransportnumber2='+ str(entity.DomesticFlightTransportnumber2)+' where OverseasFlightTransportnumber1='+ str(entity.OverseasFlightTransportnumber1))
                c.commit()
            return TwoWay(entity.OverseasFlightTransportnumber1, entity.OverseasFlightTransportnumber2, entity.DomesticFlightTransportnumber1,entity.DomesticFlightTransportnumber2)
        except:
            print("There was an error")
            return None